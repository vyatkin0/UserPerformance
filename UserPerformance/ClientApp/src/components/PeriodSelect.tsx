/**
 * Content and markup of weeks selection component
 */

import * as React from 'react';

import {ChevronLeftRounded, ChevronRightRounded} from '@material-ui/icons';
import {Divider, Link, Menu, MenuItem} from '@material-ui/core';

interface PeriodSelectProps {
    from: Date;
    to: Date;
    minYear: number;
    maxYear: number;
    onChange: Function;
}

interface PeriodSelectState {
    anchorMonthEl:HTMLElement
    anchorYearEl:HTMLElement
}

export default class PeriodSelect extends React.PureComponent<PeriodSelectProps, PeriodSelectState> {

    constructor(props){
        super(props);
        this.state={anchorMonthEl: null, anchorYearEl: null};
    }
    private setCurrentWeek = ():boolean => {
        if(this.props.from.getDay()!==1 || this.props.to.getDay()!==1)
        {
            const from = new Date(this.props.from.getFullYear(), this.props.from.getMonth(), this.props.from.getDate() - (this.props.from.getDay()===0 ? 6 : this.props.from.getDay()-1));
            const to = new Date(from.getFullYear(), from.getMonth(), from.getDate() + 7);

            this.props.onChange(from, to);
            return false;
        }

        return true;
    }

    private handleChangeMonth = (value) => {
        const startYear = this.props.to.getFullYear();
        const startDate = new Date(`${startYear}-${value}-01`);
        this.handleChangeDate(startDate);
        this.handleSelectClose();
    }

    private handleChangeYear = (value) => {
        const startMonth = this.props.to.getMonth();
        const startDate = new Date(`${value}-${startMonth+1}-01`);
        this.handleChangeDate(startDate);
        this.handleSelectClose();
    }

    private handleChangeDate = (startDate: Date) => {
        startDate.setDate(startDate.getDate()-(startDate.getDay()===0 ? 6 : startDate.getDay()-1));
        const endDate = new Date(startDate);
        endDate.setDate(endDate.getDate()+7);

        if(!isNaN(startDate.getTime()) && !isNaN(endDate.getTime())) {
            this.props.onChange(startDate, endDate);
        }
    }

    private handleNextWeek = (event: React.MouseEvent) => {
        event.preventDefault();
        if(this.setCurrentWeek()) {
            const nextDay = new Date(this.props.to);
            nextDay.setDate(nextDay.getDate() + 7);
            this.props.onChange(this.props.to, nextDay);
        }
    }

    private handlePrevWeek = (event: React.MouseEvent) => {
        event.preventDefault();
        if(this.setCurrentWeek()) {
            const prevDay = new Date(this.props.from);
            prevDay.setDate(prevDay.getDate() - 7);
            this.props.onChange(prevDay, this.props.from);
        }
    }

    private renderYears() {
        const minYear = this.props.minYear || this.props.from.getFullYear();
        const maxYear = this.props.maxYear || this.props.to.getFullYear();

        const lastDay = new Date(this.props.to);
        lastDay.setDate(lastDay.getDate() - 1);

        const selectedValue = lastDay.getFullYear();

        const result = [];
        for(let i=minYear; i<=maxYear; ++i) {
            const value = i;

            result.push(<MenuItem
                    key={i}
                    value={value}
                    selected={value===selectedValue}
                    onClick={(event) => this.handleChangeYear(value)}>
                    {value}
                </MenuItem>);
        }

        return <>
        <div
            className='period_control__select'
            onClick={this.handleYearSelectClick}
        >{lastDay.getFullYear()}</div>
        <Menu
            anchorEl={this.state.anchorYearEl}
            keepMounted
            open={Boolean(this.state.anchorYearEl)}
            onClose={this.handleSelectClose}
        >
        {result}
        </Menu>
        </>
    }

    private handleYearSelectClick = (event: React.MouseEvent<HTMLElement>) => {
        event.preventDefault();
        this.setState({anchorYearEl: event.currentTarget});
        event.currentTarget.blur();
    };

    private handleMonthSelectClick = (event: React.MouseEvent<HTMLElement>) => {
        event.preventDefault();
        this.setState({anchorMonthEl: event.currentTarget});
        event.currentTarget.blur();
    };
 
    private handleSelectClose = () => {
        this.setState({anchorMonthEl: null, anchorYearEl: null});
    };

    private renderMonths() {

        const months = [
            'January',
            'February',
            'March',
            'April',
            'May',
            'June',
            'July',
            'August',
            'September',
            'October',
            'November',
            'December',
        ];

        const lastDay = new Date(this.props.to);
        lastDay.setDate(lastDay.getDate() - 1);

        const selectedValue = lastDay.getMonth()+1;

        const result=[];
        for(let i=0; i<months.length; ++i) {
            const value = i+1;

            result.push(<MenuItem
                    key={i+1}
                    value={value}
                    selected={value===selectedValue}
                    onClick={(event) => this.handleChangeMonth(value)}>
                    {months[i]}
                </MenuItem>);
        }

        return <>
            <div
                className='period_control__select'
                onClick={this.handleMonthSelectClick}
            >{months[lastDay.getMonth()]}</div>
            <Menu
                anchorEl={this.state.anchorMonthEl}
                keepMounted
                open={Boolean(this.state.anchorMonthEl)}
                onClose={this.handleSelectClose}
            >
            {result}
            </Menu>
        </>
    }

    public render() {
        const now = new Date();
        now.setHours(0, 0, 0, 0);
        const nowNextDay = new Date(now);
        nowNextDay.setDate(nowNextDay.getDate() + 1);

        const lastDay = new Date(this.props.to);
        lastDay.setDate(lastDay.getDate() - 1);

        return <div className='period_control'>
            <Link href='#' onClick={this.handlePrevWeek}>
            <ChevronLeftRounded/></Link>
                <span>{this.props.from.getDate() + ' - ' + lastDay.getDate()}</span>
            <Link href='#' onClick={this.handleNextWeek}>
            <ChevronRightRounded/></Link>
            <Divider orientation='vertical' flexItem />
            {this.renderMonths()}
            <Divider orientation='vertical' flexItem />
            {this.renderYears()}
            <Divider orientation='vertical' flexItem />
            <div
                className='period_control__select'
                onClick={(e) => { e.preventDefault(); this.props.onChange(now, nowNextDay); e.currentTarget.blur(); }}
            >Today {now.toLocaleDateString()/*`${now.getMonth()+1}/${now.getDate()}`*/}</div>
            </div>
    }
}