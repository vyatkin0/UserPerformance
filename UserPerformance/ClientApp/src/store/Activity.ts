/**
 * Activity object interface
 */
export default interface Activity {
    id?: number; // Identifier in Database
    parentId?: number; // Parent identifier
    name: string; // Name of activity
    description: string; // Description of activity
    workCost?: number; // Spent time for activity
}