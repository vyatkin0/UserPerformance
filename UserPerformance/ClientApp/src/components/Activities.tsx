/**
 * Content and markup of Activities page
 */

import * as ActivitiesStore from '../store/Activities';
import * as React from 'react';
import * as ReactDOM from 'react-dom'

import {
    Button,
    Checkbox,
    Collapse,
    List,
    ListItem,
    ListItemIcon,
    ListItemText
} from '@material-ui/core';
import { ExpandLess, ExpandMore } from '@material-ui/icons';
import { WithStyles, withStyles } from '@material-ui/core/styles';

import Activity from '../store/Activity';
import ActivityDialog from './ActivityDialog';
import AppAlert from './AppAlert'
import { ApplicationState } from '../store';
import { connect } from 'react-redux';
import styles from '../styles';

type ActivitiesProps =  ActivitiesStore.ActivitiesState
 & WithStyles<typeof styles>
 & typeof ActivitiesStore.actionCreators;

class Activities extends React.PureComponent<ActivitiesProps> {

    public componentDidMount() {
        this.props.requestActivities();
    }

    public componentDidUpdate() {
        if(this.props.isActivityAdded && !this.props.isLoading && !this.props.isUploading) {
            this.props.requestActivities();
        }
    }

    private handleToggle = (ids:number[], checked:boolean) => {

        const selected = this.props.activities.selected.filter(id => !ids.includes(id));

        if(checked) {
            selected.push(...ids);
        }

        this.props.activities.selected = selected;  
        this.forceUpdate();
    };
    
    private handleParentToggle = (id:number, checked:boolean, children: any[]) => {
        const ids = children.map(a=>a.id);
        ids.push(id);

        if(checked) {
            this.props.activities.opened.push(id); 
        }

        this.handleToggle(ids, checked);

        this.forceUpdate();
    }

    private handleParentToggleOpen = (id:number) => {
        const opened = this.props.activities.opened.filter(i => i!==id);

        if(opened.length === this.props.activities.opened.length) {
            opened.push(id);
        }

        this.props.activities.opened = opened;
        this.forceUpdate();
    }

    private saveActivity = (id: number, name:string, description:string, workCost:number) => {
        const activity:Activity = {
            id,
            name,
            description,
            workCost,
        }

        this.props.sendActivity(activity);
    };

    public render() {
        const { classes } = this.props;
        const activities = this.props.activities;

        const groupedActivities = this.props.groupedActivities;

        const handleToggle = this.handleToggle.bind(this);
        const handleParentToggle = this.handleParentToggle.bind(this); 
        const handleParentToggleOpen =  this.handleParentToggleOpen.bind(this);

        return (
            <>
                <h1>Activities</h1> 
                {
                    Object.keys(groupedActivities).map( key => {
                        if('null' === key){
                            return null;
                        }

                        const parent = activities.activities.find(a => a.id === Number(key));
                        const parentLabelId = `checkbox-list-label-${parent.id}`;
                        const parentChecked = groupedActivities[key].every( a =>activities.selected.includes(a.id));

                        let opened = true;
                        if( parent.id && activities.opened ) {
                            opened = activities.opened.includes(parent.id);
                        }

                        return <React.Fragment key={key}>
                            <div className='flex_container'>
                                <Checkbox
                                    edge="start"
                                    checked={parentChecked}
                                    tabIndex={-1}
                                    disableRipple
                                    inputProps={{ 'aria-labelledby': parentLabelId }}
                                    onChange={()=>handleParentToggle(parent.id, !parentChecked, groupedActivities[key])}
                                />
                                <div className={'flex_container ' + classes.divParent} onClick={()=>handleParentToggleOpen(parent.id)}>
                                    <h2>{parent.name}</h2>
                                    {opened ? <ExpandLess /> : <ExpandMore />}
                                </div>
                            </div>
                            <List>
                            {groupedActivities[key].map( a =>{

                            const labelId = `checkbox-list-label-${a.id}`;
                            const checked = activities.selected.includes(a.id);
                            return <Collapse key={a.id} in={opened} timeout="auto" unmountOnExit>
                            <ListItem dense>
                            <ListItemIcon onClick={()=>handleToggle([a.id], !checked)}>
                            <Checkbox
                                className={classes.divChild}
                                edge="start"
                                checked={checked}
                                tabIndex={-1}
                                disableRipple
                                inputProps={{ 'aria-labelledby': labelId }}
                            />
                            </ListItemIcon>
                            <ListItemText
                                id={labelId}
                                className={classes.divParent}
                                primary={`${a.name} (${a.description})`}
                                onClick={(e)=>{e.preventDefault(); this.props.displayActivityDialog(true, a)}}/>
                            </ListItem>
                            </Collapse>
                            })}
                            </List>
                        </React.Fragment>
                    })
                }

                <AppAlert messages={this.props.messages}/>

                {
                    ReactDOM.createPortal(<div className={classes.divButtons}>
                        <Button variant='contained' color='primary' onClick={this.props.gotoHome}>
                        Cancel
                        </Button>
                        <Button variant='contained' color='primary' onClick={this.props.sendSelection}>
                        Save
                        </Button>
                        <Button variant='contained' color='primary' onClick={()=>this.props.displayActivityDialog(true)}>
                        Add
                        </Button>
                        </div>, document.getElementById("footerToolbar"))
                }
                
                <ActivityDialog
                    activity={this.props.openedActivity}
                    handleDelete={this.props.deleteActivity}
                    handleSave={this.saveActivity}
                    handleClose={()=>this.props.displayActivityDialog(false)}
                    open = {this.props.isDlgOpen}
                />
            </>
        );
    }
}

export default connect(
    (state: ApplicationState) => state.activities,
    ActivitiesStore.actionCreators
  )(withStyles(styles)(Activities) as any);