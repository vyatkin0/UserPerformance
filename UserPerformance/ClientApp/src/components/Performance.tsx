/**
 * Content and markup of Performance page
 */

import * as PerformanceStore from '../store/Performance';
import * as React from 'react';
import * as ReactDOM from 'react-dom'

import {
    Button,
    Popover,
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow,
    TextField
} from '@material-ui/core';
import { WithStyles, withStyles } from '@material-ui/core/styles';

import AppAlert from './AppAlert'
import { ApplicationState } from '../store';
import Paper from '@material-ui/core/Paper';
import PeriodSelect from './PeriodSelect';
import { connect } from 'react-redux';
import styles from '../styles';

type PerformanceProps =  PerformanceStore.PerformanceState
 & WithStyles<typeof styles>
 & typeof PerformanceStore.actionCreators; // ... plus action creators we've requested

class Performance extends React.PureComponent<PerformanceProps> {

    constructor(props) {
        super(props);
        this.countsInputRef = React.createRef(); // Input in popover window 
    }

    public componentDidMount() {
        
        // Support for electron environment
        const w = window as any;
        if (typeof w.ipcRenderer === 'undefined') {
            window.addEventListener('beforeunload', this.confirmExit.bind(this));
        } else {
            w.ipcRenderer.on('app-close', (event, arg) => {
                const fakeEvent = {returnValue: true, preventDefault:()=>{}};
                this.confirmExit(fakeEvent);

                event.returnValue = typeof fakeEvent.returnValue !== 'undefined';
                w.ipcRenderer.send('closed', event.returnValue);
            });
        }
 
        const now = new Date();
        const from = new Date(now.getFullYear(), now.getMonth(), now.getDate());
        const to = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1);
        this.props.requestPerformance(from, to, true);
        //this.ensureDataFetched();
    }
    
    public componentDidUpdate() {
        this.ensureDataFetched();
    }

    private ensureDataFetched() {
        this.props.requestPerformance(this.props.from, this.props.to, false);
    }

    private countsInputRef;

    private currentPopper = {
        open:false,
        anchorEl: null,
        activityId:undefined,
        dayIndex:undefined,
        countsHelper:'',
    };

    private confirmExit = (e) => {
        if(Object.keys(this.props.changes).length>0){
            // Cancel the event
            e.preventDefault(); // If you prevent default behavior in Mozilla Firefox prompt will always be shown

            // Chrome requires returnValue to be set
            e.returnValue = 'Changes are not saved!';

            return e.returnValue;
        }
        else {
            // the absence of a returnValue property on the event will guarantee the browser unload happens
            delete e['returnValue'];
        }
    }
    private handleClick = (event: React.MouseEvent<HTMLSpanElement>, activityId:number, dayIndex:number) => {
        if( this.props.performance.currentDay < 0) {
            return;
        }
        
        const offset = this.props.performance.currentDay-dayIndex;
        if( offset<0 || offset>=this.props.performance.editedDays ) {
            return;
        }

        this.currentPopper = {
            open: true,
            anchorEl: event.currentTarget,
            activityId,
            dayIndex,
            countsHelper:'',
        };

        this.forceUpdate();
    };

    private handleClose = (event, update: boolean) => {
        event.preventDefault();

        const {activityId, dayIndex} = this.currentPopper;
        
        if ( typeof activityId === 'undefined'
            || activityId === null) return;

        if(update) {
            if(0 === activityId) {
                update = this.props.performance.days[dayIndex].hours != this.countsInputRef.current.value;
            } else {
                const userActivity = this.props.performance.userActivities.find(ea => ea.activity.id === activityId);

                update = userActivity && (userActivity.counts[dayIndex] != this.countsInputRef.current.value);
            }

            if(update && !this.validateFields(this.countsInputRef.current.value)) {
                this.forceUpdate();
                return;
            }

            this.props.updateActivityCount(activityId, dayIndex, this.countsInputRef.current.value);
        }

        this.currentPopper = {
            open: false,
            anchorEl: null,
            activityId:undefined,
            dayIndex:undefined,
            countsHelper:'',
        };

        this.forceUpdate();
    };
    
    private validateFields = (counts:string) => {

        const wrongValue = 'Wrong value';

        const n = Number(counts.replace(',','.'));
        let result = isFinite(n);

        if(result) {
            if(this.currentPopper.activityId===0) {
                if(n>24) {
                    result = false;
                    this.currentPopper.countsHelper = 'More than 24 hours per day';
                }
            }

            if (n<0) {
                result = false;
                this.currentPopper.countsHelper = wrongValue;
            }
        }
        else {
            // Для отработанных часов в неделю canExpand===false
            if(this.currentPopper.activityId===0) {
                const allowedChars = ['V','H'];
                result = allowedChars.includes(counts.trim().toUpperCase());
            }

            if(!result) {
                this.currentPopper.countsHelper = wrongValue;
            }
        }

        return result;
    }

    private handlePopperKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
        switch(event.key)
        {
        case 'Enter':
            event.preventDefault();
            this.handleClose(event, true);
            break;
        }
    }

    private renderStub = () => 
        <div className='flex_container'>
            <div>
                <h1>Activities management</h1>
            </div>
        </div> 


    /**
     * Content and markup small windows for edit counts of task and day hours
     */
    private renderEditPopover = () => {

        const { classes } = this.props;
        const currentPopper = this.currentPopper;

        return <Popover
            id='simplePopper'
            open={currentPopper.open}
            anchorEl={currentPopper.anchorEl}
            onClose={(e, reason) => this.handleClose(e, reason==='backdropClick')}
            anchorOrigin={{
                vertical: 'top',
                horizontal: 'left',
            }}
            transformOrigin={{
                vertical: 'top',
                horizontal: 'left',
            }}>
            <div className={classes.popover}>
                {currentPopper.anchorEl && <TextField variant='outlined'
                    autoFocus
                    inputRef={this.countsInputRef}
                    label='Value'
                    defaultValue={currentPopper.anchorEl.innerText}
                    margin='dense'
                    InputProps={{
                        classes: {
                            input: classes.textField,
                        }}}
                    error={Boolean(currentPopper.countsHelper)}
                    helperText = {currentPopper.countsHelper}
                    onKeyPress =  {this.handlePopperKeyPress}
                />}
                {Boolean(currentPopper.countsHelper) && <div><Button autoFocus variant='outlined' size='small' onClick={(e) =>this.handleClose(e, false)}>
                    Close
                </Button></div>}
            </div>
            </Popover>
    }

    /**
     * Content and markup of summary for month performance
     */
    private renderMonthInfo = () => {
        const performance = this.props.performance;

        const now = new Date();
        now.setHours(0, 0, 0, 0);
        const nowNextDay = new Date(now);
        nowNextDay.setDate(nowNextDay.getDate() + 1);
        
        const lastDay = new Date(this.props.to);
        lastDay.setDate(lastDay.getDate() - 1);
        
        return <>
        <div className='flex_container'>
            <div>
                <h1>Activities management</h1>
                <h3>{performance.userName}</h3>
            </div>
            <PeriodSelect
            from = {this.props.from}
            to = {this.props.to}
            minYear = {this.props.performance.minYear}
            maxYear = {this.props.performance.maxYear}
            onChange = {(from: Date, to: Date)=> this.props.requestPerformance(from, to, false)}
            />
        </div>
        <div className='performance_container'>
            Month performance
            <span className={(performance.monthPerformance < 100 ? 'warning ' : '') + 'performance_data'} >{performance.monthPerformance+'%'}</span>
            Working days
            <span className='performance_data'>{performance.monthWorkDays}</span>
        </div>
        </>
    }

    /**
     * Content and markup of table with performance for each day of specified period
     */
    private renderDaysInfo = () => {

        const { classes } = this.props;

        const performance = this.props.performance;
        const handleClick = this.handleClick.bind(this);
        
        return <TableContainer component={Paper}>
            <Table className={classes.table} size='small' aria-label='simple table'>
                <TableHead>
                <TableRow>
                    <TableCell align='left'>Name</TableCell>
                    <TableCell className={classes.monthHeadCell} align='center'></TableCell>
                    {performance.days.map((day,index) =>
                        <TableCell key={'month:'+ index} className={classes.numCell} align='center'>{`${day.day.getDate()}.${day.day.getMonth()+1}`}</TableCell>)}
                </TableRow>
                </TableHead>
                <TableBody>
                <TableRow>
                    <TableCell component='th' scope='row'>Day performance</TableCell>
                    <TableCell align='center'></TableCell>
                    {performance.days.map((d, index)=>{

                        let monthPerformance = '';
                        const dayHours = Number(d.hours);

                        let className = null;

                        if(dayHours>0)
                        { 
                            const mp = performance.userActivities.reduce((result, a)=> {
                                if(a.counts[index]>0 && a.activity.workCost>0) {
                                    return result + a.counts[index]*a.activity.workCost/60/dayHours;
                                }
                                
                                return result;
                            } , 0);

                            monthPerformance = parseFloat((mp * 100).toFixed(2)) + '%'

                            if(mp<1) {
                                className = classes.warnCell;
                            }
                        }

                        return <TableCell key={'month:'+ index} align='center' className={className}>{monthPerformance}</TableCell>})}
                </TableRow>
                <TableRow>
                    <TableCell component='th' scope='row'>Spent hours</TableCell>
                    <TableCell align='center'></TableCell>
                    {performance.days.map((d, index)=> {
                        let className = index<=performance.currentDay && index+performance.editedDays>performance.currentDay ? classes.cellSpan + ' ' + classes.cellSpanCurrent : classes.cellSpan;

                        if(this.isValueChanged(d.day.valueOf(), 0)) {
                            className += ' ' + classes.changedValue;
                        }

                        return <TableCell key={'month:'+ index} align='center'>
                        {<span className={className} onClick={(e)=>handleClick(e, 0, index)}>{d.hours}</span>}
                    </TableCell>})}
                </TableRow>
                </TableBody>
            </Table>
        </TableContainer>
    }

    /**
     * Footer content and markup
     */
    private renderFooterPortal = () => {
        const { classes } = this.props;
        
        const saveBtnDisabled = !(Object.keys(this.props.changes).length>0);

        return ReactDOM.createPortal(<div className={classes.divButtons}>
            <Button variant='contained' color='secondary' disabled={saveBtnDisabled} onClick={this.props.sendPerformance}>
            Save
            </Button>
            <Button variant='contained' color='primary' href='#/activities'/*this.props.gotoActivities*/>
            Select activities
            </Button>
            </div>, document.getElementById('footerToolbar'))
    }

    /**
     * User Performance content and markup
     */
    private renderUserActivities = () => {

        const performance = this.props.performance;

        const { classes } = this.props;

        const groupedActivities = performance ? performance.groupedActivities : {};

        const handleClick = this.handleClick.bind(this);

        return Object.keys(groupedActivities).map(key => {
            if ('null' === key) {
                return null;
            }

            const parent = performance.userActivities.find(ua => ua.activity.id === Number(key)).activity;

            return <React.Fragment key={key}>
                <h2>{parent.name}</h2>
                <TableContainer component={Paper}>
                    <Table className={classes.table} size='small' aria-label='simple table'>
                        <TableHead>
                            <TableRow>
                                <TableCell align='left'>Activity</TableCell>
                                <TableCell className={classes.monthHeadCell} align='center'>Spent time</TableCell>
                                <TableCell className={classes.monthHeadCell} align='center'>Count per month</TableCell>
                                {performance.days.map((day, index) =>
                                    <TableCell key={key + ':' + index} className={classes.numCell} align='center'>{`${day.day.getDate()}.${day.day.getMonth() + 1}`}</TableCell>)}
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {groupedActivities[key].map(ua => <TableRow key={ua.activity.id}>
                                <TableCell component='th' scope='row'>{ua.activity.name}</TableCell>
                                <TableCell align='center'>{ua.activity.workCost}</TableCell>
                                <TableCell align='center'>{ua.countPerMonth}</TableCell>
                                {performance.days.map((d, index) => {
                                    let className = index <= performance.currentDay && index + performance.editedDays > performance.currentDay ? classes.cellSpan + ' ' + classes.cellSpanCurrent : classes.cellSpan;

                                    if (this.isValueChanged(d.day.valueOf(), ua.activity.id)) {
                                        className += ' ' + classes.changedValue;
                                    }

                                    return <TableCell key={ua.activity.id + ':' + index} align='center'>
                                        {<span className={className} onClick={(e) => handleClick(e, ua.activity.id, index)}>{ua.counts[index]}</span>}
                                    </TableCell>
                                })}
                            </TableRow>)}
                        </TableBody>
                    </Table>
                </TableContainer>
            </React.Fragment>
        })
    }

    isValueChanged(day, activityId) :boolean {
        if (Object.keys(this.props.changes).length<1) {
            return false;
        }

        return this.props.changes[day]
            && this.props.changes[day][activityId];
    }
    
    public render() {
        const performance = this.props.performance;
        
        if(!performance || !performance.days || performance.days.length<1) {
            return this.renderStub();
        }

        return (
            <>
                {this.renderMonthInfo()}
                {this.renderEditPopover()}
                {this.renderDaysInfo()}
                {this.renderUserActivities()}
 
                {this.props.messages.length>0 && <AppAlert messages={this.props.messages}/>}

                { this.renderFooterPortal() }
            </>
        );
    }
}

export default connect(
    (state: ApplicationState) => state.performance,
    PerformanceStore.actionCreators
  )(withStyles(styles)(Performance) as any);