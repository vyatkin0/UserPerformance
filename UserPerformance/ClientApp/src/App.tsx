import * as React from 'react';

import Activities from './components/Activities';
import Performance from './components/Performance';
import { Route } from 'react-router';

export default () =>
    <>
        <Route exact path='/activities' component={Activities} />
        <Route exact path='/' component={Performance} />
    </>
