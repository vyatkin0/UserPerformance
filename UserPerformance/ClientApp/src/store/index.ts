/**
 * Application store
 */

import * as Activities from './Activities';
import * as Performance from './Performance';

// The top-level state object
export interface ApplicationState {
    performance: Performance.PerformanceState | undefined;
    activities: Activities.ActivitiesState | undefined;
}

// Whenever an action is dispatched, Redux will update each top-level application state property using
// the reducer with the matching name. It's important that the names match exactly, and that the reducer
// acts on the corresponding ApplicationState property type.
export const reducers = {
    performance: Performance.reducer,
    activities: Activities.reducer
};

// This type can be used as a hint on action creators so that its 'dispatch' and 'getState' params are
// correctly typed to match your store.
export interface AppThunkAction<TAction> {
    (dispatch: (action: TAction) => void, getState: () => ApplicationState): void;
}
