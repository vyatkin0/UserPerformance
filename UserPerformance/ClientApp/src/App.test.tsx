import * as ActivitiesStore from './store/Activities';
import * as PerformanceStore from './store/Performance';
import * as React from 'react';
import * as ReactDOM from 'react-dom';

import Activities from './components/Activities';
import App from './App';
import { MemoryRouter } from 'react-router-dom';
import Performance from './components/Performance';
import PeriodSelect from './components/PeriodSelect';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

describe('Renders component without crashing', () => {
    it('App renders without crashing', () => {
        const storeFake = (state: any) => ({
            default: () => {},
            subscribe: () => {},
            dispatch: () => {},
            getState: () => ({ ...state })
        });

        const state = {
            activities: ActivitiesStore.emptyState(),
            performance: PerformanceStore.emptyState(),
        };

        const store = storeFake(state) as any;
        
        const elem = document.createElement('div');
        elem.id = 'footerToolbar';

        document.body.appendChild(elem);

        ReactDOM.render(
            <Provider store={store}>
                <MemoryRouter>
                    <Activities/>
                    <Performance/>
                    <App/>
                </MemoryRouter>
            </Provider>, document.createElement('div'));
    });


    it('PeriodSelect renders without crashing', () => {
    const div = document.createElement('div');
    const from = new Date(1);
    const to = new Date(4000);

    ReactDOM.render(<PeriodSelect
        from = {from}
        to = {to}
        minYear = {from.getFullYear()}
        maxYear = {to.getFullYear()}
        onChange = {(from: Date, to: Date)=> ({from,to})}
        />, div);
    });
})

describe('Component snapshots', () => {
    beforeAll(() => {
        ReactDOM.createPortal = jest.fn((element, node) => {
        return element;
      })
    })
  
    afterEach(() => {
        ReactDOM.createPortal.mockClear();
    })
    it('App should render correctly with Node or Function', () => {
       const storeFake = (state: any) => ({
            default: () => {},
            subscribe: () => {},
            dispatch: () => {},
            getState: () => ({ ...state })
        });

        const state = {
            activities: ActivitiesStore.emptyState(),
        };

        const store = storeFake(state) as any;

        const component = renderer.create(<Provider store={store}>
            <MemoryRouter>
                <Activities/>
            </MemoryRouter>
        </Provider>);

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
  });