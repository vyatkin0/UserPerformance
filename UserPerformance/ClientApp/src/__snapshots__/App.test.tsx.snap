// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshots App should render correctly with Node or Function 1`] = `
<h1>
  Activities
</h1>
`;
